### PR TITLE
Simplify Poly multiplication, speed up interpolation.

### PR DESCRIPTION
**Pull request by afck**
This reduces time for interpolation by about 20% on my machine.

Also add more `Mul` implementations.

### CODE REVIEW

File: src/poly.rs [hunk 1]
```diff
@@ -211,22 +211,19 @@ impl<'a, B: Borrow<Poly>> ops::Mul<B> for &'a Poly {
...
-                }
-                c
-            }).collect();
-
-        match Poly::new(coeff) {
-            Ok(poly) => poly,
-            Err(e) => panic!("Failed to create a new `Poly` duing muliplication: {}", e),
+        let rhs = rhs.borrow();
+        let mut coeff = vec![Fr::zero(); self.coeff.len() + rhs.borrow().coeff.len() - 1];
+        let mut s = SecretKey::default(); // Temporary value will be zeroed on drop.
```
Comment (DrPeterVanNostrand) [inline src/poly.rs]: This is definitely an abuse of `SecretKey`, I agree that a `Secret` type (issue #24) should be added for situations such as these.

Comment (afck) [reply src/poly.rs]: Agreed. I removed it and added back the TODO.

**Comment by afck** [comment]
I added another commit which speeds up interpolation by an additional ~45% on my machine. Division is expensive!

**Comment by afck** [comment]
And another one, which gives between 25% (degree 5) and 84% (degree 40) speedup.

**Pull request merged**